### PR TITLE
[#477] Typesafe Zatoshi APIs for amounts

### PR DESCRIPTION
**Pull request by ccjernigan**
When implementing this, I started considering whether we could maintain the existing APIs and mark them as deprecated, then adding additional new methods that leverage Zatoshi.  However there are some places where this wouldn't be viable such as the WalletBalance object.  So I went ahead and decided we should just do the breaking API changes now.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/tests) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the demo app and try the changes?
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [x] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [x] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [x] Run the app: Did you run the demo app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Review by pacu — COMMENTED** [review]
made some non-blocking comments

File: demo-app/src/main/java/cash/z/ecc/android/sdk/demoapp/demos/listtransactions/TransactionViewHolder.kt [hunk 1]
```diff
@@ -23,7 +24,7 @@ class TransactionViewHolder<T : ConfirmedTransaction>(itemView: View) : Recycler
 
     fun bindTo(transaction: T?) {
         val isInbound = transaction?.toAddress.isNullOrEmpty()
-        amountText.text = transaction?.value.convertZatoshiToZecString()
+        amountText.text = transaction?.valueInZatoshi.convertZatoshiToZecString()
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply demo-app/src/main/java/cash/z/ecc/android/sdk/demoapp/demos/listtransactions/TransactionViewHolder.kt:27]: Transaction is a Room DAO so the type is long (which can’t be changed) and it requires an extension function to convert to the Zatoshi object.

File: MIGRATIONS.md [hunk 2]
```diff
@@ -1,6 +1,14 @@
 Troubleshooting Migrations
 ==========
 
+Upcoming Migration to Version 1.7 from 1.6
+--------------------------------------
+Various APIs used `Long` value to represent Zatoshi currency amounts.  Those APIs now use a typesafe `Zatoshi` class.  When passing amounts, simply wrap Long values with the Zatoshi constructor `Zatoshi(Long)`.  When receiving values, simply unwrap Long values with `Zatishi.value`.
+
+`WalletBalance` no longer has uninitialized default values.  This means that `Synchronizer` fields that expose a WalletBalance now use `null` to signal an uninitialized value.  Specifically this means `Synchronizer.orchardBalances`, `Synchronzier.saplingBalances`, and `Synchronizer.transparentBalances` have nullable values now.
+
+`ZcashSdk.ZATOSHI_PER_ZEC` has been moved to `Zatishi.ZATOSHI_PER_ZEC`.
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply MIGRATIONS.md:10]: Thanks 😂

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Zatoshi.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * A unit of currency used throughout the SDK.
+ *
+ * End users (e.g. app users) generally are not shown Zatoshi values.  Instead they are presented
+ * with ZEC, which is a decimal value represented only as a String.  ZEC are not used internally,
+ * to avoid floating point imprecision.
+ */
+data class Zatoshi(val value: Long) {
+    init {
+        require(value >= MIN_INCLUSIVE) { "Zatoshi must be in the range [0, $MAX_INCLUSIVE]" }
```
Comment (HonzaR) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Zatoshi.kt:12]: No big deal, but I'd suggest to change the failure message to _"Zatoshi must be in the range [$MIN_INCLUSIVE, $MAX_INCLUSIVE]"_.

**Review by HonzaR — APPROVED** [review]
Good improvement, I like to have it in Zatoshi, as we already have it the same way in the wallet project. I placed one minor suggestion. The next thing is that four of our `sdk-lib` tests are failing locally. They're all related to birthday height. Are we aware of that?

**Pull request merged**